### ISSUE TITLE
Create a confluence_children macro

### ISSUE THREAD

**Issue by michitux** [xwikisas/xwiki-pro-macros#220]
The standard children macro in XWiki doesn't support the page parameter that exists in Confluence. A new `confluence_children` macro should be created to cover this use case.

Main macro code:

```
{{velocity}}
#set ($root = "document:$!wikimacro.parameters.page")
{{documentTree root="$services.rendering.escape($root, 'xwiki/2.1')" /}}
{{/velocity}
```

An optional parameter `page` of type `org.xwiki.model.reference.DocumentReference` should be added for the page parameter.

**Comment by michitux** [comment]
Documentation of the corresponding Confluence macro: https://confluence.atlassian.com/doc/children-display-macro-139501.html

**Comment by trrenty** [comment]
Do we only need to cover the page parameter of the confluence macro? Or are we interested in covering the other parameters as well. There are few parameters that can be easily converted to a `documentTree` parameter. 

Also, there seems to be different types of values for the `page` parameter. Are we interested in supporting all of them? 

> Specify the page to display children for, from either the current space or a different space. Enter:
> 
>     '/' — to list the top-level pages of the current space, i.e. those without parents.
>     'pagename' — to list the children of the specified page.
>     'spacekey:' — to list the top-level pages of the specified space.
>     'spacekey:pagename' — to list the children of the specified page in the specified space.

**Comment by michitux** [comment]
Here the XML source of that macro call:

```xml
<p><ac:structured-macro ac:name="children" ac:schema-version="2" ac:macro-id="23ba62b6-8087-4c5d-b514-38403d205c84"><ac:parameter ac:name="depth">1</ac:parameter><ac:parameter ac:name="style">h3</ac:parameter><ac:parameter ac:name="page"><ac:link><ri:page ri:content-title="IT-Anleitungen/Instructions" /></ac:link></ac:parameter></ac:structured-macro></p
```

**Comment by michitux** [comment]
> Do we only need to cover the page parameter of the confluence macro? Or are we interested in covering the other parameters as well. There are few parameters that can be easily converted to a `documentTree` parameter.

The more parameters the better I guess, in the case I saw the `page` parameter was the important one. It depends on how much time and effort we/you want to put into this.

**Comment by vmassol** [comment]
> The standard children macro in XWiki doesn't support the page parameter that exists in Confluence. A new confluence_children macro should be created to cover this use case.

@michitux If there's a need to improve the children macro from XWiki, it's also interesting if it adds something useful. I'd find it interesting to be able to have an option "parent" parameter for the default children macro, and I'd find it a better solution than creating a new macro which makes it harder for user who'll see both macros which is confusing. I'd bring it as a proposal on the forum.

**Comment by michitux** [comment]
> @michitux If there's a need to improve the children macro from XWiki, it's also interesting if it adds something useful. I'd find it interesting to be able to have an option "parent" parameter for the default children macro, and I'd find it a better solution than creating a new macro which makes it harder for user who'll see both macros which is confusing. I'd bring it as a proposal on the forum.

Note that XWiki itself has the [Document Tree Macro](https://extensions.xwiki.org/xwiki/bin/view/Extension/Document%20Tree%20Macro) that supports all these parameters - it's just not really user-friendly (e.g., there is no picker for the root and there is no distinction between simpler and more advanced parameters). In my opinion it would make more sense to make the document tree macro easier to use than to duplicate parts of its functionality in the children macro.

The problem here is that we need to support the exact parameters Confluence uses, we have no choice to, e.g., name the parameter `parent`, it must be `page`. However, we can easily hide the `confluence_children` macro from the macro dialog to avoid user confusion.

**Comment by vmassol** [comment]
>  it's just not really user-friendly (e.g., there is no picker for the root and there is no distinction between simpler and more advanced parameters). In my opinion it would make more sense to make the document tree macro easier to use than to duplicate parts of its functionality in the children macro.

Sure, that looks good to me. 

> The problem here is that we need to support the exact parameters Confluence uses, we have no choice to, e.g., name the parameter parent, it must be page

I'm surprised because it basically means we can almost never use any existing xwiki macros when converting from confluence. That doesn't seem a good solution to have to rewrite macros just to map parameter names. Is it not possible in the confluence filter stream code to do this mapping somewhere?

> However, we can easily hide the confluence_children macro from the macro dialog to avoid user confusion.

I guess it would need to be the default children macro that should be hidden since the confluence one would be used.

**Comment by michitux** [comment]
> That doesn't seem a good solution to have to rewrite macros just to map parameter names. Is it not possible in the confluence filter stream code to do this mapping somewhere?

As far as I'm aware this is possible, but this means that the mapping needs to exist and be supported at the time of the migration. On the other hand, with the approach to import the macro as-is, it is possible to add better support after the import has been completed. That way, the support for the macro can be improved while users are already working with the imported content.

> I guess it would need to be the default children macro that should be hidden since the confluence one would be used.

Both is possible, also through configuration in the XWiki installation so it can be decided on a case-by-case basis. My assumption here was that while existing content from Confluence would continue to use the migrated macro, new content would use a more standard XWiki macro like the document tree macro.

**Comment by vmassol** [comment]
> As far as I'm aware this is possible, but this means that the mapping needs to exist and be supported at the time of the migration. On the other hand, with the approach to import the macro as-is, it is possible to add better support after the import has been completed. That way, the support for the macro can be improved while users are already working with the imported content.

ok, then this should be fixed as it's painful indeed. The generic filter steam component could introduce a concept of macro provider which would be in charge of deciding what macro to call and with which parameter. It would be a component with a hint being the confluence macro id (for example), and the component implementation would be provided by some extension (ie not in XS).

WDYT?

**Comment by michitux** [comment]
> then this should be fixed

The Confluence migrator already provides the possibility to convert macros during import. This is for example used by the [task manager](https://github.com/xwikisas/application-task/blob/9d8ab3938234e27b342a5408a6659ffcb07302e5/application-task-default/src/main/java/com/xwiki/task/internal/TaskMacroConverter.java#L42). It is my understanding that it is an explicit decision of the pro macros team not to use this mechanism but to instead import most of the content as-is and provide bridge macros because, as I mentioned already, this provides the possibility to improve compatibility with the imported content after the import has been completed. In most cases, wikis that are migrated from Confluence are in active use and as the content is edited after import, it is not possible to repeat the import after we've improved support for macros in the Confluence content.

There are other possibilities of course, we could also provide a way to basically execute macro converters after the import has been completed, but this is a possibility we currently don't provide. If you think this should be changed, please discuss this internally in XWiki SAS with the people who are responsible for this.

**Comment by trrenty** [comment]
Note that there already is an issue on Confluence Migrator that proposes the same solution https://github.com/xwikisas/application-confluence-migrator-pro/issues/16

**Comment by trrenty** [comment]
I created a basic version of the `confluence_children` macro. Additionally to the `page` parameter I also took into consideration the `first` parameter that maps nicely to the `limit` parameter of the `documentTree` macro. I also opted to not display the attachments and translations to align better with the confluence macro.

If we want to support the sorting and styling parameters, we should create a separate macro that is even more in line with the behaviour of  the confluence macro.
